### DISCUSSION TITLE
how do we use this

### DISCUSSION THREAD

**Discussion by xalteropsx** [huggingface/diffusers#8570] [Q&A] [answered]
@tolgacangoz bro  how do we use  custom_pipeline="lpw_stable_diffusion"  https://github.com/huggingface/diffusers/discussions/6976 with compel   // i do have custom lpw but i want to know if diffuser have builtin with compel support 

this is also make me think if it not possible https://github.com/huggingface/diffusers/issues/3171  but i saw recently some update on lpw pipeline no wonder if it work or not

**Comment by xalteropsx** [comment]
@tolgacangoz  nvm notice here https://github.com/damian0815/compel/issues/54

**Comment by asomoza** [comment, marked as answer]
Don't use a mix of them, they do exactly the same thing, the lpw pipeline has its own prompt weighting which is what compel does, you don't need to use compel if you use the lpw pipeline.

The lpw pipelines use the automatic1111 style of weighting like this:

`a cat with a (red:1.5) ball`

where compel uses it like this:

`a cat with a red++ ball`

**Reply by xalteropsx** [reply]
ahh i got it but what if 1.5 is 1.8 or 1.9 something ? i see i was doing it little wrong not sure but i will show u my approach but i pretty unsure what these people do in order to make it work

**Reply by asomoza** [reply]
Is based in automatic1111 prompt weighting, you can learn how to use it [here](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#attentionemphasis)

**Reply by xalteropsx** [reply]
i mean i want to learn compel better  give me a few 1hr i will be back from outside and show my code

**Reply by asomoza** [reply]
oh compel is easier, just add more `+` for each emphasis you want (or `-`). But I remember that it doesn't support long prompts for SDXL or it has problems with it.

**Reply by xalteropsx** [reply]
but they have floating value upper is little more control over + - is not idea how much it increase or decrease bit confusing 

equivalent to 1.1, ++ is 1.1^2, +++ is 1.1^3, etc; - means 0.9, -- means 0.9^2, etc.
is little weird in auto11 people could do more control 
sec i send u their code i using also same approach it was on their github

**Reply by asomoza** [reply]
yeah, the lpw approach is a lot more easy to understand and to fine tune. I don't really like the `+` and `-` but compel has something similar to the float values, it's just a little harder to do and understand:

```python
("a cat with a", "red", "ball").and(1, 1.5, 1)
```

**Reply by xalteropsx** [reply]
nvm wrong file  i miss read just above but it was lark used let me research it

**Reply by xalteropsx** [reply]
@asomoza https://github.com/vladmandic/automatic 

is the compel parser design to take input as compel ++ or an an paranthese value of lpw ?

**Reply by xalteropsx** [reply]
i using compel just to exceed 77 limit is lpw also exceed 77 limit ? bit of confused as new toward it

**Reply by asomoza** [reply]
I don't really know about which one sd.next supports, it uses both, the original automatic1111 backend and a diffusers backend, so maybe both depending on the backend used? That file you posted loads compel but also parses the prompt with the lpw format.

I don't know how compel does it, but I'm guessing it splits the embeddings after the tokenizer, so yeah, for both use cases the tokenizer will throw a warning. Just keep in mind that the custom pipelines aren't working when you use `from_single_file`

**Reply by xalteropsx** [reply]
i guess i understand thanks for clearing it >.< /  i was missused it by putting lpw format directly onto it lol and parse it using compel guess good to ask and help to fix problem seems like it all depend on regex , i will try manipulate it later with my own format

**Reply by asomoza** [reply]
I mean that this code doesn't load the `lpw_stable_diffusion_xl`:

```python
pipe = StableDiffusionXLPipeline.from_single_file(
    model_path,
    torch_dtype=torch.float16,
    custom_pipeline="lpw_stable_diffusion_xl",
    use_safetensors=True,
)
```

[Here](https://github.com/huggingface/diffusers/issues/7666#issuecomment-2168722449) I posted a simple test you can do to see if it's working or not.

Compel is not a pipeline, it's a library so it doesn't have this problem.

**Reply by xalteropsx** [reply]
tested u right

**Reply by xalteropsx** [reply]
may be i download an model from_pretrained to test lpw thanks alot bro 
if so the case then look like lpw more dominant then compel but that make really hard my whole thing depend .fromsinglefile lol >.< sorry for troubling but really awesome help

**Reply by asomoza** [reply]
yeah, the lpw pipelines do support more than 77 tokens, the `lpw` means `long prompt weighting` but this doesn't work with `from_single_file` and the `custom_pipeline` argument. You have to either download the pipeline locally and use it directly or with `from_pretrained`.

If you need to use it with a single file checkpoint, as I suggested you can [save it locally](https://raw.githubusercontent.com/huggingface/diffusers/main/examples/community/lpw_stable_diffusion_xl.py) and import it like this:

```python
from lpw_stable_diffusion_xl import SDXLLongPromptWeightingPipeline

pipe = SDXLLongPromptWeightingPipeline.from_single_file(
    model_path,
    torch_dtype=torch.float16,
    use_safetensors=True,
)
```

and no problem, glad to help.

**Reply by xalteropsx** [reply]
ah i havent aware of this thing lol this really rock do have sd one as well ?

**Reply by asomoza** [reply]
[yep](https://raw.githubusercontent.com/huggingface/diffusers/main/examples/community/lpw_stable_diffusion.py)

**Reply by xalteropsx** [reply]
awesome bro it really save me alot of time to switching from single and downloading model again thanks u million may god bless u >.< once i succeed hope to donate something
